src/rust/engine/options/src/env.rs

```diff
@@ -14,13 +15,42 @@ pub struct Env {
   pub(crate) env: HashMap<String, String>,
 }
 
+#[derive(Debug)]
+pub struct DroppedEnvVars {
+  pub non_utf8_keys: Vec<OsString>,
+  pub keys_with_non_utf8_values: Vec<String>,
+}
+
 impl Env {
   pub fn new(env: HashMap<String, String>) -> Self {
     Self { env }
   }
 
-  pub fn capture() -> Self {
-    Self::new(env::vars().collect::<HashMap<_, _>>())
+  pub fn capture_lossy() -> (Self, DroppedEnvVars) {
+    Self::do_capture_lossy(env::vars_os())
+  }
+
+  pub(crate) fn do_capture_lossy<I>(env_os: I) -> (Self, DroppedEnvVars)
+  where
+    I: Iterator<Item = (OsString, OsString)>,
+  {
+    let mut env: HashMap<String, String> = HashMap::with_capacity(env_os.size_hint().0);
+    let mut dropped = DroppedEnvVars {
+      non_utf8_keys: Vec::new(),
+      keys_with_non_utf8_values: Vec::new(),
+    };
+    for (os_key, os_val) in env_os {
+      match (os_key.into_string(), os_val.into_string()) {
+        (Ok(key), Ok(val)) => {
+          env.insert(key, val);
+        }
+        (Ok(key), Err(_)) => dropped.keys_with_non_utf8_values.push(key),
+        (Err(os_key), _) => dropped.non_utf8_keys.push(os_key),
+      }
+    }
+    dropped.non_utf8_keys.sort();
```

Comment:
I think any form of display considerations being mixed here is not a great idea. The OsString -> String conversion for display is now gone, which I think is good, but I'd take it 1 step further is you actually agree with the reasoning here. It either all applies or it doesn't it seems to me.

Reply:
Do you mean to not sort here? I'd prefer to keep it. It's potentially useful to the caller (and ensures consistent output in tests, so it's not just a display consideration), and at worst does no harm. If the caller wants to sort they have to make a mut copy and sort it, which is a couple of extra lines, and more allocation, for each vec. Seems unnecessary.

Reply:
Your last sentence is untrue in terms of copying. For example, this works and prints `a z`:
```rust
fn main() {
    pub struct Foo {
        pub bar: Vec<String>
    }
    let f = Foo { bar: vec!["z".into(), "a".into()] };
    let mut b = f.bar;
    b.sort();
    println!("{bar}", bar = b.join(" "));
}
```

I find it strange to do a 3rd party sort. The sort isn't built into the contract of the struct right now; so some are sorted, some not. If your claims about sorting being needed for test reasons, then that's obviously valid. It doesn't ring true to me, but your head is in the code.

Reply:
So... a couple of extra lines, as I said... True, no extra allocation if you're ok moving the Vec, and making the struct subsequently unusable. Which I guess we are in our current usage of this struct.

Still, since I don't really feel strongly about it, I will remove the sorts here and push them to the caller.

Reply:
Excellent. Of the 3, caller decides, struct guarantees, 3rdparty does it I think the way you had it was the worst possible.

Reply:
> and making the struct subsequently unusable.

You take the example too literally. You can take the whole struct as mut in the reciever and mutate one of its field in-place.

Reply:
A more on-point example that works:
```rust
fn main() {
    pub struct Foo {
        pub bar: Vec<String>
    }
    fn example() -> (u64, Foo) {
        (42, Foo { bar: vec!["z".into(), "a".into()] })
    }
    let (mol, mut f) = example();
    f.bar.sort();
    println!("{mol}: {bar}", bar = f.bar.join(" "));
}
```